File: room.go
//go:build wasip1 || tinygo.wasm

package gamekit

import (
	"context"
	"math/rand"
	"time"
)

// Room is the authoring surface: local reads answered from the cached
// CallContext (zero host calls), effects via host functions. A Room handle is
// valid only inside the callback that received it.
type Room interface {
	// Local reads.
	Members() []Player
	Has(p Player) bool
	Count() int
	Config() RoomConfig
	Rand() *rand.Rand
	Now() time.Time
	Settled() bool

	// Effects (host calls).
	Send(p Player, f *Frame)
	Identical(f *Frame)
	SetInputContext(ctx InputContext)
	End(res Result)
	Post(res Result)
	Log(msg string)

	Services() Services
}

// Handler is the game's per-room behavior — the lean wasm surface (OnWake is
// the host heartbeat; there are no ticks, timers, or frame callbacks).
type Handler interface {
	OnStart(r Room)
	OnJoin(r Room, p Player)
	OnLeave(r Room, p Player)
	OnInput(r Room, p Player, in Input)
	OnWake(r Room)
	OnClose(r Room)
}

// Base supplies no-op defaults so a game overrides only what it needs.
type Base struct{}

func (Base) OnStart(Room)                {}
func (Base) OnJoin(Room, Player)         {}
func (Base) OnLeave(Room, Player)        {}
func (Base) OnInput(Room, Player, Input) {}
func (Base) OnWake(Room)                 {}
func (Base) OnClose(Room)                {}

// Game is the module entry: static metadata plus the room behavior factory.
type Game interface {
	Meta() GameMeta
	NewRoom(cfg RoomConfig, svc Services) Handler
}

// ---- Services ----------------------------------------------------------------

// KVStore is the durable per-user KV, already namespaced to this game and one
// account (the HOST derives both — a guest can never name another namespace).
// Signatures mirror the native sdk so KV patterns port unchanged; the context
// is accepted and ignored (the host bounds the call).
type KVStore interface {
	Get(ctx context.Context, key string) ([]byte, bool, error)
	Set(ctx context.Context, key string, value []byte, rule MergeRule) error
	Delete(ctx context.Context, key string) error
}

// Account is a live, account-scoped handle for a Player.
type Account interface {
	ID() string
	Handle() string
	Kind() Kind
	Store() KVStore
}

// AccountStore yields an Account for a Player.
type AccountStore interface{ For(p Player) Account }

// ConfigStore is the slug-bound, read-only per-game config surface.
type ConfigStore interface {
	Get(ctx context.Context, key string) ([]byte, bool, error)
}

// Services is the ABI v1 service bundle (no chat, no spectate).
type Services struct {
	Accounts AccountStore
	Config   ConfigStore
}

// ---- implementation -----------------------------------------------------------

// room is the live Room implementation, refreshed per callback from the
// decoded CallContext.
type room struct {
	ctx callContext
	rng *rand.Rand
}

func (r *room) Members() []Player  { return r.ctx.members }
func (r *room) Count() int         { return len(r.ctx.members) }
func (r *room) Config() RoomConfig { return r.ctx.cfg }
func (r *room) Rand() *rand.Rand   { return r.rng }
func (r *room) Now() time.Time     { return time.Unix(0, r.ctx.nowUnixNanos) }
func (r *room) Settled() bool      { return r.ctx.settled }

func (r *room) Has(p Player) bool {
	for _, m := range r.ctx.members {
		if m == p {
			return true
		}
	}
	return false
}

func (r *room) index(p Player) int {
	for i, m := range r.ctx.members {
		if m == p {
			return i
		}
	}
	return -1
}

func (r *room) Send(p Player, f *Frame) {
	idx := r.index(p)
	if idx < 0 || f == nil {
		return
	}
	m := alloc(encodeFrame(f))
	hostSend(uint64(idx), m.Offset())
	m.Free()
}

func (r *room) Identical(f *Frame) {
	if f == nil {
		return
	}
	m := alloc(encodeFrame(f))
	hostIdentical(m.Offset())
	m.Free()
}

func (r *room) SetInputContext(ctx InputContext) { hostSetInputContext(uint64(ctx)) }

func (r *room) End(res Result) {
	m := alloc(encodeResult(res, r.ctx.members))
	hostEnd(m.Offset())
	m.Free()
}

func (r *room) Post(res Result) {
	m := alloc(encodeResult(res, r.ctx.members))
	hostPost(m.Offset())
	m.Free()
}

func (r *room) Log(msg string) {
	m := allocStr(msg)
	hostLog(1, m.Offset())
	m.Free()
}

func (r *room) Services() Services {
	return Services{Accounts: accountStore{r}, Config: configStore{}}
}

type accountStore struct{ r *room }

func (s accountStore) For(p Player) Account {
	idx := s.r.index(p)
	if idx < 0 {
		// Departed player delivered by the host as the final roster entry
		// (the leave callback): resolve by account id.
		for i, m := range s.r.ctx.members {
			if m.AccountID == p.AccountID {
				idx = i
				break
			}
		}
	}
	if idx < 0 {
		return nil
	}
	return account{idx: idx, p: s.r.ctx.members[idx]}
}

type account struct {
	idx int
	p   Player
}

func (a account) ID() string     { return a.p.AccountID }
func (a account) Handle() string { return a.p.Handle }
func (a account) Kind() Kind     { return a.p.Kind }
func (a account) Store() KVStore { return kvStore{idx: a.idx} }

type kvStore struct{ idx int }

func (k kvStore) Get(_ context.Context, key string) ([]byte, bool, error) {
	km := allocStr(key)
	off := hostKVGet(uint64(k.idx), km.Offset())
	km.Free()
	v, ok := readBytesFree(off)
	return v, ok, nil
}

func (k kvStore) Set(_ context.Context, key string, value []byte, rule MergeRule) error {
	km, vm, rm := allocStr(key), alloc(value), allocStr(string(rule))
	hostKVSet(uint64(k.idx), km.Offset(), vm.Offset(), rm.Offset())
	km.Free()
	vm.Free()
	rm.Free()
	return nil
}

func (k kvStore) Delete(_ context.Context, key string) error {
	km := allocStr(key)
	hostKVDelete(uint64(k.idx), km.Offset())
	km.Free()
	return nil
}

type configStore struct{}

func (configStore) Get(_ context.Context, key string) ([]byte, bool, error) {
	km := allocStr(key)
	off := hostConfigGet(km.Offset())
	km.Free()
	v, ok := readBytesFree(off)
	return v, ok, nil
}

File: rust/src/room.rs
//! The `Room` handle: local reads answered from the cached CallContext (zero
//! host calls), effects via host functions with the SDK-owned frame-diffing
//! discipline. A `Room` is valid only inside the callback that received it.

use std::cell::RefCell;

use crate::broadcast::{SdkState, BROADCAST_SLOT};
use crate::frame::Frame;
use crate::host;
use crate::input::InputContext;
use crate::rng::SplitMix64;
use crate::types::{MergeRule, Outcome, Player, RoomConfig};
use crate::wire::{encode_outcome, CallCtx};

thread_local! {
    /// The per-instance SDK state surviving across callbacks: baseline table,
    /// PRNG, roster fingerprint. Fully safe (see `SdkState` docs).
    pub(crate) static STATE: RefCell<SdkState> = RefCell::new(SdkState::new());
}

/// The authoring surface handed to every [`crate::Handler`] callback.
pub struct Room {
    pub(crate) ctx: CallCtx,
}

impl Room {
    // ---- local reads (zero host calls) ----------------------------------------

    /// The current roster. On `on_leave`, the departed player is the FINAL
    /// entry (no longer a member, present for that callback only).
    pub fn members(&self) -> &[Player] {
        &self.ctx.members
    }

    /// Whether `p` is in the current roster.
    pub fn has(&self, p: &Player) -> bool {
        self.ctx.members.contains(p)
    }

    /// Roster size (see [`Room::members`] for the `on_leave` caveat).
    pub fn count(&self) -> usize {
        self.ctx.members.len()
    }

    pub fn config(&self) -> &RoomConfig {
        &self.ctx.cfg
    }

    /// The room clock in unix nanoseconds — monotonic non-decreasing, the only
    /// time a game may use (never wall time; hibernation depends on it).
    pub fn now_unix_nanos(&self) -> i64 {
        self.ctx.now_unix_nanos
    }

    /// Whether the room has already settled (`end` was called).
    pub fn settled(&self) -> bool {
        self.ctx.settled
    }

    /// Next value from the room PRNG (seeded from the room seed; state lives in
    /// linear memory so hibernation replays the exact sequence).
    pub fn rng_u64(&mut self) -> u64 {
        STATE.with(|s| {
            let mut st = s.borrow_mut();
            let seed = self.ctx.cfg.seed;
            st.rng.get_or_insert_with(|| SplitMix64::new(seed)).next_u64()
        })
    }

    /// Uniform value in `0..n` from the room PRNG (`0` when `n == 0`).
    pub fn rng_below(&mut self, n: u64) -> u64 {
        STATE.with(|s| {
            let mut st = s.borrow_mut();
            let seed = self.ctx.cfg.seed;
            st.rng.get_or_insert_with(|| SplitMix64::new(seed)).below(n)
        })
    }

    // ---- effects (host calls) ---------------------------------------------------

    /// Deliver a frame to one player. The SDK diffs against that player slot's
    /// baseline, keyframes on bootstrap/roster change, mirrors the
    /// host-returned epoch, and retries a rejected delta as a keyframe in-call
    /// — no render is ever lost. A departed/unknown player is a no-op.
    pub fn send(&mut self, p: &Player, f: &Frame) {
        let Some(idx) = self.index(p) else {
            return;
        };
        STATE.with(|s| s.borrow_mut().send_slot(idx, f));
    }

    /// Broadcast one frame to every player (host `identical`). On accept the
    /// SDK reconciles EVERY per-player baseline to this frame, so a later
    /// [`Room::send`] diffs correctly.
    pub fn identical(&mut self, f: &Frame) {
        STATE.with(|s| s.borrow_mut().send_slot(BROADCAST_SLOT, f));
    }

    /// Select how subsequent input resolves (Nav/Command/Text). Host-side
    /// state: survives hibernation; do not re-issue on resume.
    pub fn set_input_context(&mut self, ctx: InputContext) {
        host::host_set_input_context(ctx as i64);
    }

    /// Settle the room exactly once with the final rankings.
    pub fn end(&mut self, res: &Outcome) {
        host::host_end(&encode_outcome(res, &self.ctx.members));
    }

    /// Record a leaderboard result without ending the room.
    pub fn post(&mut self, res: &Outcome) {
        host::host_post(&encode_outcome(res, &self.ctx.members));
    }

    /// Host-side info log (also where panics surface — the default panic hook
    /// prints to virtualized stderr, which lands in the room log).
    pub fn log(&mut self, msg: &str) {
        host::host_log(1, msg);
    }

    // ---- durable per-user KV + read-only config ----------------------------------

    /// Durable per-user KV read; `None` = not found. Namespacing is host-side:
    /// this game + this account, derived from the roster index — a guest can
    /// never name another namespace.
    pub fn kv_get(&self, p: &Player, key: &str) -> Option<Vec<u8>> {
        let idx = self.kv_index(p)?;
        host::host_kv_get(idx, key)
    }

    /// Durable per-user KV write with the account-merge `rule`.
    pub fn kv_set(&mut self, p: &Player, key: &str, value: &[u8], rule: MergeRule) {
        let Some(idx) = self.kv_index(p) else {
            return;
        };
        host::host_kv_set(idx, key, value, rule.as_str());
    }

    /// Durable per-user KV delete.
    pub fn kv_delete(&mut self, p: &Player, key: &str) {
        let Some(idx) = self.kv_index(p) else {
            return;
        };
        host::host_kv_delete(idx, key);
    }

    /// Read-only per-game config; `None` = not found.
    pub fn config_get(&self, key: &str) -> Option<Vec<u8>> {
        host::host_config_get(key)
    }

    // ---- internals ---------------------------------------------------------------

    fn index(&self, p: &Player) -> Option<usize> {
        self.ctx.members.iter().position(|m| m == p)
    }

    /// KV resolves a departed player (delivered as the final roster entry on
    /// `on_leave`) by account id when exact equality misses — mirroring the Go
    /// SDK, so "save on leave" works.
    fn kv_index(&self, p: &Player) -> Option<usize> {
        self.index(p)
            .or_else(|| self.ctx.members.iter().position(|m| m.account_id == p.account_id))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::host::{reset_test_host, with_test_host};
    use crate::types::Kind;
    use crate::wire::CallCtx;

    fn room_with(members: Vec<Player>) -> Room {
        Room {
            ctx: CallCtx {
                now_unix_nanos: 1,
                cfg: RoomConfig { seed: 99, ..Default::default() },
                members: std::rc::Rc::new(members),
                settled: false,
                roster_epoch: None,
                roster_unchanged: false,
            },
        }
    }

    fn player(id: &str) -> Player {
        Player { account_id: id.into(), handle: id.into(), conn: "c".into(), kind: Kind::Member }
    }

    #[test]
    fn send_to_unknown_player_is_a_noop() {
        reset_test_host();
        STATE.with(|s| *s.borrow_mut() = SdkState::new());
        let mut r = room_with(vec![player("a")]);
        r.send(&player("ghost"), &Frame::new());
        with_test_host(|h| assert!(h.sent.is_empty()));
        r.send(&player("a"), &Frame::new());
        with_test_host(|h| assert_eq!(h.sent.len(), 1));
    }

    #[test]
    fn rng_is_seeded_from_room_seed_and_persists() {
        reset_test_host();
        STATE.with(|s| *s.borrow_mut() = SdkState::new());
        let mut r = room_with(vec![]);
        let a = r.rng_u64();
        let b = r.rng_u64();
        assert_ne!(a, b, "sequence advances");
        let mut reference = crate::rng::SplitMix64::new(99);
        assert_eq!(a, reference.next_u64(), "seeded from ctx seed");
    }

    #[test]
    fn kv_resolves_departed_player_by_account_id() {
        reset_test_host();
        let mut r = room_with(vec![player("a"), player("b")]);
        // A clone whose handle changed (e.g. departed snapshot) still resolves.
        let mut departed = player("b");
        departed.handle = "old-handle".into();
        r.kv_set(&departed, "wins", b"3", MergeRule::Sum);
        with_test_host(|h| {
            assert_eq!(h.kv.get(&(1, "wins".to_string())).map(|v| v.as_slice()), Some(&b"3"[..]));
        });
        assert_eq!(r.kv_get(&player("b"), "wins"), Some(b"3".to_vec()));
        assert_eq!(r.kv_get(&player("nope"), "wins"), None);
    }

    #[test]
    fn outcome_and_log_reach_the_host() {
        reset_test_host();
        let mut r = room_with(vec![player("a")]);
        r.log("hello");
        r.end(&Outcome::default());
        r.set_input_context(InputContext::Text);
        with_test_host(|h| {
            assert_eq!(h.logs, vec![(1, "hello".to_string())]);
            assert_eq!(h.ends.len(), 1);
            assert_eq!(h.input_contexts, vec![2]);
        });
    }
}
